shared/lib/src/bloc_complex/catalog/catalog_bloc.dart

```diff
@@ -0,0 +1,135 @@
+import 'dart:async';
+import 'dart:math';
+
+import 'package:flutter/widgets.dart';
+import 'package:reactive_exploration/common/models/product.dart';
+import 'package:reactive_exploration/src/bloc_complex/catalog/catalog_page.dart';
+import 'package:reactive_exploration/src/bloc_complex/catalog/catalog_slice.dart';
+import 'package:rxdart/rxdart.dart';
+
+/// This component encapsulates the logic of fetching products from
+/// a database, page by page, according to position in an infinite list.
+///
+/// Only the data that are close to the current location are cached, the rest
+/// is thrown away.
+class CatalogBloc {
+  static const _productsPerPage = 10;
+
+  /// We're using ReactiveX's [PublishSubject] here because we want to easily
+  /// buffer the stream. See [CatalogBloc] constructor.
+  final _indexController = PublishSubject<int>();
+
+  /// These are the pages stored in memory. For O(1) retrieval, we're storing
+  /// them in a [Map]. The key value is [CatalogPage.startIndex].
+  final _pages = <int, CatalogPage>{};
+
+  /// A set of pages that are currently being fetched from the network.
+  /// They are identified by their [CatalogPage.startIndex].
+  final _pagesBeingRequested = Set<int>();
+
+  final _sliceSubject = BehaviorSubject<CatalogSlice>();
+
+  CatalogBloc() {
+    _indexController.stream
+        // Don't try to update too frequently.
+        .bufferTime(const Duration(milliseconds: 500))
+        // Don't update when there is no need.
+        .where((batch) => batch.isNotEmpty)
+        .listen(_handleIndexes);
+  }
+
+  /// An input of the indexes that the [ListView.builder]
+  /// (or [GridView.builder]) is getting in its builder callbacks. Just push
+  /// the index that you get in a [IndexedWidgetBuilder] down this [Sink].
+  ///
+  /// The component uses this input to figure out which pages it should
+  /// be requesting from the network.
+  Sink<int> get index => _indexController.sink;
+
+  /// The currently available data, as a slice of the (potentially infinite)
+  /// catalog.
+  Stream<CatalogSlice> get slice => _sliceSubject.stream;
+
+  /// Outputs the [CatalogPage.startIndex] given an arbitrary index of
+  /// a product.
+  int _getPageStartFromIndex(int index) =>
+      (index ~/ _productsPerPage) * _productsPerPage;
+
+  /// This will handle the incoming [indexes] (that were requested by
+  /// a [IndexedWidgetBuilder]) and, if needed, will fetch missing data.
+  void _handleIndexes(List<int> indexes) {
+    final int minIndex = indexes.fold(0x7fffffff, min);
```

Comment:
Are you using the hex value as a maxint of sorts? If so, I'd throw a comment above this line or put it into a constant just for the readability of it.